File: townJudge.js
/**
 * @param {number} N
 * @param {number[][]} trust
 * @return {number}
 */
var findJudge = function(N, trust) {
    //base
    if(trust == null) return -1;
    let indegrees = new Array(N).fill(0);
    for(let t of trust){
        //decrease the corresponding indegree value of the node of the trustee node
        indegrees[t[0]-1]--;
        //increase the corresponding indegree value of the node of the trusted node
        indegrees[t[1]-1]++;
    }
    
    for(let i = 0; i< indegrees.length; i++){
        if(indegrees[i] == N-1){
            return i+1;
        }
    }
    
    return -1;
    
};

//timeComplexity : O(E(edges)-no of entries in the trust array) + O(N)
//space complexity : O(N)

File: townjudge.java
class Solution {
    public int findJudge(int n, int[][] trust) {
        //TC-O(V+E),SC-O(n)
        // create a graph with entries of trust as edges and track count of incoming edges increase by 1 for every incming edge and decrease one for utgoing array.If any count=n-1,that is the answer
        int[] indegrees = new int[n];
        for(int[] tr:trust){
            indegrees[tr[0]-1]--;
            indegrees[tr[1]-1]++;
        }
        for(int i=0;i<n;i++){
            if(indegrees[i]==n-1){
                return i+1;
            }
        }
        return -1;
    }
}